src/main/java/bio/terra/service/dataset/flight/inheritsteward/AdjustStewardMembersStep.java

```diff
@@ -0,0 +1,61 @@
+package bio.terra.service.dataset.flight.inheritsteward;
+
+import bio.terra.common.iam.AuthenticatedUserRequest;
+import bio.terra.model.PolicyModel;
+import bio.terra.service.auth.iam.IamResourceType;
+import bio.terra.service.auth.iam.IamRole;
+import bio.terra.service.auth.iam.IamService;
+import bio.terra.service.dataset.flight.DatasetWorkingMapKeys;
+import bio.terra.service.job.JobMapKeys;
+import bio.terra.stairway.FlightContext;
+import bio.terra.stairway.FlightMap;
+import bio.terra.stairway.Step;
+import bio.terra.stairway.StepResult;
+import bio.terra.stairway.exception.RetryException;
+import java.util.List;
+import java.util.Objects;
+import java.util.UUID;
+
+public record AdjustStewardMembersStep(
+    AuthenticatedUserRequest userReq, IamService iamService, boolean inheritSteward)
+    implements Step {
+
+  interface AddRemoveApi {
+    PolicyModel addRemoveMember(
+        AuthenticatedUserRequest userReq,
+        IamResourceType iamResourceType,
+        UUID resourceId,
+        String policyName,
+        String userEmail);
+  }
+
+  private void addRemoveStewardMembers(FlightContext context, boolean inheritSteward) {
+    FlightMap workingMap = context.getWorkingMap();
+    List<UUID> snapshotIds =
+        Objects.requireNonNull(workingMap.get(DatasetWorkingMapKeys.SNAPSHOT_IDS, List.class));
+    FlightMap inputParams = context.getInputParameters();
+    List<String> custodians =
+        Objects.requireNonNull(
+            inputParams.get(JobMapKeys.CUSTODIAN_USERS.getKeyName(), List.class));
+    AddRemoveApi api =
+        inheritSteward ? iamService::addPolicyMember : iamService::deletePolicyMember;
```

Comment:
If we are at all concerned about performance, we may want to add special `IamService` APIs to handle this migration. Each call to these APIs makes two Sam calls; one to change the policy, and another to retrieve the current policy state. Also note that each API call results in a new journal entry, although the journal is per-snapshot so that doesn't seem too noisy.

Reply:
> Each call to these APIs makes two Sam calls; one to change the policy, and another to retrieve the current policy state which is unneeded here.

Looking at the code, it would be reasonable to make the callers who need the policy to explicitly call Sam separately from this API.

Reply:
I have a PR onto this branch with this change. Or it could be merged in as a separate change after this PR is merged. https://github.com/DataBiosphere/jade-data-repo/pull/1939